### PR TITLE
Computing Epsilon/Delta for aggregation

### PR DESCRIPTION
**Pull request by dvadym**

### CODE REVIEW

**Review by miracvbasaran — CHANGES_REQUESTED** [review]

File: pipeline_dp/budget_accounting.py [hunk 1]
```diff
@@ -146,6 +169,64 @@ def scope(self, weight: float):
...
+                          self._total_delta * budget_ratio)
+        # No expectations on aggregations, no way to compute budget.
+        return None
+
+    def _check_aggregation_restrictions(self):
+        if self._expected_n_aggregations:
+            actual_n_aggregations = len(self._actual_aggregation_weights)
+            if actual_n_aggregations != self._expected_n_aggregations:
+                raise ValueError(
+                    f"'n_aggregations' in the constructor of BudgetAccountant is "
```
[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/budget_accounting.py]: Thanks, I've updated this and other error messages.

No rules, I've just tried to keep line length <= 80 characters, and I put newline as earlier as possible :)

File: pipeline_dp/budget_accounting.py [hunk 2]
```diff
@@ -201,22 +282,30 @@ def _normalise_mechanism_weights(self):
...
 
         Args:
             total_epsilon: epsilon for the entire pipeline.
             total_delta: delta for the entire pipeline.
+            n_aggregations: number of aggregations for which 'self' manages
+              the budget. All aggregations should have budget_weight = 1.
+              If None, any number of aggregations is allowed.
+            aggregation_weights: weights of aggregations for which 'self'
+              manages the budget. If None, any number of aggregations and
+              weights are allowed.
```
[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/budget_accounting.py]: I've updated the docstring. Please check

**Review by dvadym — COMMENTED** [review]
Thank for comments! PTAL

File: pipeline_dp/budget_accounting.py [hunk 3]
```diff
@@ -285,29 +291,32 @@ class NaiveBudgetAccountant(BudgetAccountant):
...
+             which  'self' manages the budget. All aggregations should have
+             'budget_weight' = 1. When specified all aggregations will have
+             equal budget. It is useful to ensure that the pipeline has fixed
+             number of DP aggregations.
+            aggregation_weights: 'budget_weight' of aggregations for which
+             'self' manages the budget. It is useful to ensure that the pipeline
+              has a fixed number of DP aggregations with fixed weights.
+
+        if num_aggregations and aggregation_weights are not set, there are no
+        restrictions on number of aggregations nor their budget_weights
```
Comment (miracvbasaran) [inline pipeline_dp/budget_accounting.py:311]: nit: "on *the* number of aggregations"

nit: "budget weights" instead of "budget_weights"?

**Review by miracvbasaran — CHANGES_REQUESTED** [review]

File: examples/movie_view_ratings/run_without_frameworks.py [hunk 4]
```diff
@@ -33,11 +33,8 @@ def main(unused_argv):
...
 
     # Define the privacy budget available for our computation.
-    budget_accountant = pipeline_dp.NaiveBudgetAccountant(
-        total_epsilon=1,
-        total_delta=1e-6,
-        n_aggregations=1,
-    )
+    budget_accountant = pipeline_dp.NaiveBudgetAccountant(total_epsilon=1,
+                                                          total_delta=1e-6)
     # aggregation_weights=[2])
```
[the remark this answers is not in the document]
Comment (dvadym) [reply examples/movie_view_ratings/run_without_frameworks.py]: Thanks, done

**Review by dvadym — COMMENTED** [review]
PTAL

File: pipeline_dp/budget_accounting.py [hunk 5]
```diff
@@ -300,15 +300,15 @@ def __init__(self,
...
              which  'self' manages the budget. All aggregations should have
-             'budget_weight' = 1. When specified all aggregations will have
+             'budget_weight' = 1. When specified, all aggregations will have
              equal budget. It is useful to ensure that the pipeline has fixed
              number of DP aggregations.
             aggregation_weights: 'budget_weight' of aggregations for which
              'self' manages the budget. It is useful to ensure that the pipeline
               has a fixed number of DP aggregations with fixed weights.
 
         if num_aggregations and aggregation_weights are not set, there are no
```
Comment (miracvbasaran) [inline pipeline_dp/budget_accounting.py:310]: We need to make the same change for PLD Accountant below.

Also, I see many references to `budget_weight` in this file (for example a couple lines above). Maybe they could all be `budget weights` or `aggregation weights`?

Comment (dvadym) [reply pipeline_dp/budget_accounting.py]: Thanks, PLDAccountant is udpated

`budget_weight` is the name for the [field](https://github.com/OpenMined/PipelineDP/blob/784961601d26afa0a08beedc7fe15587c3da9987/pipeline_dp/aggregate_params.py#L79) in `AggregateParams`

**Review by miracvbasaran — APPROVED** [review]

**Pull request merged**